### ISSUE TITLE
Add Temurin to `JvmVendorSpec`

### ISSUE THREAD

**Issue by msgilligan** [gradle/gradle#18004]
Add **Temurin** to `JvmVendorSpec`. Given that Temurin effectively replaces `AdoptOpenJDK`, there needs to be a way to specify that Temurin is the desired vendor.

**Comment by mlopatkin** [comment]
Seems like Temurin builds aren't consistent in their vendor naming. The JDK 8 builds have `java.vendor = Temurin` but JDK 11 and 16 have `java.vendor = Eclipse Foundation`. There is an [issue](https://github.com/adoptium/temurin-build/issues/2664) about it opened in the Adoptium project, so I'd like to postpone adding a constant for Temurin until it is resolved.

In the meantime it is possible to specify the vendor with a string-based check:
```gradle
java {
    toolchain {
        languageVersion.set(JavaLanguageVersion.of(11))
        vendor.set(JvmVendorSpec.matching("Eclipse"))
    }
}
```
or 
```gradle
java {
    toolchain {
        languageVersion.set(JavaLanguageVersion.of(8))
        vendor.set(JvmVendorSpec.matching("Temurin"))
    }
}
```

**Comment by evpaassen** [comment]
> Seems like Temurin builds aren't consistent in their vendor naming. The JDK 8 builds have `java.vendor = Temurin` but JDK 11 and 16 have `java.vendor = Eclipse Foundation`. There is an [issue](https://github.com/adoptium/temurin-build/issues/2664) about it opened in the Adoptium project, so I'd like to postpone adding a constant for Temurin until it is resolved.

I think this issue might be resolved now at the Adoptium side with adoptium/temurin-build#2705.

Furthermore, [the Gradle docs say auto-provisioning of JDKs happens with AdoptOpenJDK](https://docs.gradle.org/current/userguide/toolchains.html#sec:provisioning). How will Adoptium/Temurin affect that, as all future releases will come from that project instead? Will Gradle use Adoptium/Temurin for auto-provisioning, or will we be stuck with an old AdoptOpenJDK version?

**Comment by mlopatkin** [comment]
>I think this issue might be resolved now at the Adoptium side with adoptium/temurin-build#2705

We'd probably support all three in the end.

>Furthermore, the Gradle docs say auto-provisioning of JDKs happens with AdoptOpenJDK. How will Adoptium/Temurin affect that, as all future releases will come from that project instead?

So far the underlying AdoptOpenJDK API provides links to fresh Temurin (for Hotspot) and IBM Semeru (for OpenJ9) binaries, so Gradle just picks up that (unless `vendor = ADOPTOPENJDK` is specified, this is currently broken). The Adoptium project which maintains the API plans to provide builds via the old API for a while but will stop publishing newer builds eventually. The new Adoptium's API has no OpenJ9 builds so completely switching to it right now would be a regression. There are plans to address this in some newer versions of Gradle though I cannot promise an exact timeline.

**Comment by evpaassen** [comment]
> So far the underlying AdoptOpenJDK API provides links to fresh Temurin (for Hotspot) and IBM Semeru (for OpenJ9) binaries, so Gradle just picks up that (unless `vendor = ADOPTOPENJDK` is specified, this is currently broken). The Adoptium project which maintains the API plans to provide builds via the old API for a while but will stop publishing newer builds eventually. The new Adoptium's API has no OpenJ9 builds so completely switching to it right now would be a regression. There are plans to address this in some newer versions of Gradle though I cannot promise an exact timeline.

That makes sense. Thank you for the clarification.

**Comment by cvmocanu** [comment]
Why was this ticket closed? I'm using Gradle 8.4 (not 7.4), and `Temurin` is nowhere to be found in `JvmVendorSpec`.

What is the current recommended way to force Gradle to pick Java 21 from Eclipse Temurin?

**Comment by mlopatkin** [comment]
> Why was this ticket closed? I'm using Gradle 8.4 (not 7.4), and `Temurin` is nowhere to be found in `JvmVendorSpec`.
> 
> What is the current recommended way to force Gradle to pick Java 21 from Eclipse Temurin?

You can use [ADOPTIUM](https://docs.gradle.org/current/javadoc/org/gradle/jvm/toolchain/JvmVendorSpec.html#ADOPTIUM) as a vendor constant for the Temurin JDK.
